### PR TITLE
KNOX-2859 - Token Management UI improvements

### PR DESCRIPTION
**Pull request by smolnar82**
## What changes were proposed in this pull request?

Replaced the old Angular2 Datatable with a more modern Material Table implementation and configured filtering, sorting, and pagination on both tables on the Token Management UI.

## How was this patch tested?

Manual testing with a huge amount of tokens created (2k) to ensure the new implementation still performs.

EnhancedTokenManagementUI

### CODE REVIEW

File: knox-token-management-ui/token-management/app/token.management.component.ts [hunk 1]
```diff
@@ -57,13 +117,29 @@ export class TokenManagementComponent implements OnInit {
...
     }
 
     fetchKnoxTokens(impersonated: boolean): void {
         this.tokenManagementService.getKnoxTokens(this.userName, impersonated)
-            .then(tokens => impersonated ? this.doAsKnoxTokens = tokens : this.knoxTokens = tokens);
+            .then(tokens => this.populateTokens(impersonated, tokens));
+    }
+
+    populateTokens(impersonated: boolean, tokens: KnoxToken[]) {
+	    if (impersonated) {
```
Comment (zeroflag) [inline knox-token-management-ui/token-management/app/token.management.component.ts:130]: nit: missing indentation

File: knox-token-management-ui/token-management/app/token.management.component.ts [hunk 2]
```diff
@@ -42,6 +55,53 @@ export class TokenManagementComponent implements OnInit {
...
+
+        this.knoxTokens.sortingDataAccessor = (item, property) => {
+           switch(property) {
+             case 'metadata.comment': return item.metadata.comment;
+             default: return item[property];
+           }
+        };
+
+        this.doAsKnoxTokens.sortingDataAccessor = (item, property) => {
+	       let normalizedPropertyName = property.replace('impersonation.', '');
```
Comment (zeroflag) [inline knox-token-management-ui/token-management/app/token.management.component.ts:98]: nit: wrong indentation

File: knox-token-management-ui/token-management/app/token.management.component.ts (hunk 1 above)
Comment (smolnar82) [reply knox-token-management-ui/token-management/app/token.management.component.ts]: Fixed.

File: knox-token-management-ui/token-management/app/token.management.component.ts (hunk 2 above)
Comment (smolnar82) [reply knox-token-management-ui/token-management/app/token.management.component.ts]: Fixed.

**Review by zeroflag — APPROVED** [review]

File: knox-token-management-ui/token-management/app/token.management.component.ts [hunk 3]
```diff
@@ -57,13 +117,29 @@ export class TokenManagementComponent implements OnInit {
     }
 
     fetchAllKnoxTokens(): void {
-        this.fetchKnoxTokens(true);
         this.fetchKnoxTokens(false);
+        this.fetchKnoxTokens(true);
```
Comment (pzampino) [inline knox-token-management-ui/token-management/app/token.management.component.ts:121]: Is the order important here? Just trying to understand the change.

File: knox-token-management-ui/token-management/app/token.management.component.html [hunk 4]
```diff
@@ -20,90 +20,116 @@
...
-            </tbody>
-		    <tfoot>
-		    <tr>
-		        <td colspan="6">
-		            <mfBootstrapPaginator [rowsOnPageSet]="[5,10,15]"></mfBootstrapPaginator>
-		        </td>
-		    </tr>
-		    </tfoot>
-        </table>
+    <div class="table-responsive" style="width:100%; overflow: auto; overflow-y: scroll; padding: 10px 0px 0px 0px;">
```
Comment (pzampino) [inline knox-token-management-ui/token-management/app/token.management.component.html:83]: Why are tokens acquired with impersonation presented in a distinct table from "normal" tokens? Couldn't it be a single table? The relationship between the tables isn't entirely clear to me, and you can only act on rows in the My Knox Tokens table.

**Review by pzampino — COMMENTED** [review]
This is a really nice improvement, especially if it performs as well with high token volume as your test have indicated. I just have some questions.

File: knox-token-management-ui/token-management/app/token.management.component.ts (hunk 3 above)
Comment (smolnar82) [reply knox-token-management-ui/token-management/app/token.management.component.ts:121]: Nope, the order is irrelevant. I just made to change to reflect the order of the tables on te GUI: my tokens first then the impersonation tokens.

File: knox-token-management-ui/token-management/app/token.management.component.html (hunk 4 above)
Comment (smolnar82) [reply knox-token-management-ui/token-management/app/token.management.component.html:83]: In theory, they could be, but I made this decision when token impersonation was introduced (so this is not a new thing here). This is easier to understand and the UX is better this way (IMO, at least)
The relationship is straightforward:
- in the first table we list all tokens generated by the logged-in user for himself/herself: `userName=logged-in user`
- in the second one we list all tokens generated by the logged-in user on behalf of other users: `username=otherUser; createdBy=logged-in user`

The same is described [here](https://knox.apache.org/books/knox-2-0-0/user-guide.html#Token+Management).

Comment (pzampino) [reply knox-token-management-ui/token-management/app/token.management.component.html:83]: I recognize that this is not directly related to this PR.
I think my confusion may be due to the inability to disable/revoke the tokens in the second table, but that is a separate topic.

**Review by pzampino — APPROVED** [review]
LGTM

**Pull request merged**